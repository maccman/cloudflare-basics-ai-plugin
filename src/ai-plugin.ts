import { RouteOptions, Router, json } from 'cloudflare-basics'
import { AiPluginOptions, AiPlugin } from './types'

export function buildAiPlugin(options: AiPluginOptions): AiPlugin {
  const aiPlugin: AiPlugin = {
    schema_version: 'v1',
    auth: options.auth
      ? options.auth.type === 'oauth'
        ? {
            type: 'oauth',
            client_url: options.auth.clientUrl,
            scope: options.auth.scope,
            authorization_url: options.auth.authorizationUrl,
            authorization_content_type: 'application/json',
            verification_tokens: {
              openai: options.auth.openaiVerificationToken,
            },
          }
        : options.auth.type === 'service_http'
        ? {
            type: 'service_http',
            authorization_type: options.auth.authorizationType,
          }
        : options.auth.type === 'user_http'
        ? {
            type: 'user_http',
            authorization_type: options.auth.authorizationType,
          }
        : {
            type: 'none',
          }
      : {
          type: 'none',
        },
    name_for_human: options.humanName,
    name_for_model: options.modelName,
    description_for_human: options.humanDescription,
    description_for_model: options.modelDescription,
    contact_email: options.contactEmail,
    legal_info_url: options.legalInfoUrl,
    logo_url: options.logoUrl,
    api: {
      type: 'openapi',
      has_user_authentication: false,
      url: options.apiUrl,
    },
  }

  return aiPlugin
}

export function AiPluginRoute<Env>(options: Omit<AiPluginOptions, 'apiUrl'>) {
  return async (route: RouteOptions<Env>) => {
    const origin = route.request.url.origin
    const apiUrl = `${origin}/openapi.json`

    return json(buildAiPlugin({ ...options, apiUrl }))
  }
}

export function addAiPluginRoute<Env>(
  router: Router<Env>,
  options: Omit<AiPluginOptions, 'apiUrl'>,
  path = '/.well-known/ai-plugin.json'
) {
  router.get(path, AiPluginRoute(options))
}
