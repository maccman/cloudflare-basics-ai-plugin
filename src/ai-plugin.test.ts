import { describe, expect, it } from 'vitest'
import { buildAiPlugin } from './ai-plugin'

describe('ai-plugin', () => {
  it('should generate an ai-plugin spec', () => {
    const aiPlugin = buildAiPlugin({
      auth: {
        type: 'oauth',
        clientUrl: 'https://example.com/oauth/authorize',
        scope: 'read write',
        authorizationUrl: 'https://example.com/oauth/token',
        openAiVerificationToken: '123',
      },
      humanName: 'Reflect Notes',
      modelName: 'reflect-notes',
      humanDescription:
        'A plugin that allows the user to save notes. For example, saving a summary of their ChatGPT conversation history.',
      modelDescription:
        'Reflect notes plugin for ChatGPT. This plugin allows the user to save notes. For example, saving a summary of their ChatGPT conversation history.',
      contactEmail: 'foo@example.com',
      legalInfoUrl: 'https://example.com/legal',
      logoUrl: 'https://logo.clearbit.com/reflect.app',
      apiUrl: 'https://example.com/openapi.json',
    })

    expect(aiPlugin).toMatchSnapshot()
  })
})
