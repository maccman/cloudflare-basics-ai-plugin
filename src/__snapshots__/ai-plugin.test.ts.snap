// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ai-plugin > should generate an ai-plugin spec 1`] = `
{
  "api": {
    "has_user_authentication": false,
    "type": "openapi",
    "url": "https://example.com/openapi.json",
  },
  "auth": {
    "authorization_content_type": "application/json",
    "authorization_url": "https://example.com/oauth/token",
    "client_url": "https://example.com/oauth/authorize",
    "scope": "read write",
    "type": "oauth",
    "verification_tokens": {
      "openai": "123",
    },
  },
  "contact_email": "foo@example.com",
  "description_for_human": "A plugin that allows the user to save notes. For example, saving a summary of their ChatGPT conversation history.",
  "description_for_model": "Reflect notes plugin for ChatGPT. This plugin allows the user to save notes. For example, saving a summary of their ChatGPT conversation history.",
  "legal_info_url": "https://example.com/legal",
  "logo_url": "https://logo.clearbit.com/reflect.app",
  "name_for_human": "Reflect Notes",
  "name_for_model": "reflect-notes",
  "schema_version": "v1",
}
`;
